lib/logstash/filters/geoip.rb

```diff
@@ -58,9 +64,24 @@ class LogStash::Filters::GeoIP < LogStash::Filters::Base
   # is still valid GeoJSON.
   config :target, :validate => :string, :default => 'geoip'
 
+  # GeoIP lookup is surprisingly expensive. This filter uses an LRU cache to take advantage of the fact that
+  # IPs agents are often found adjacent to one another in log files and rarely have a random distribution.
+  # The higher you set this the more likely an item is to be in the cache and the faster this filter will run.
+  # However, if you set this too high you can use more memory than desired.
+  #
+  # Experiment with different values for this option to find the best performance for your dataset.
+  #
+  # This MUST be set to a value > 0. There is really no reason to not want this behavior, the overhead is minimal
+  # and the speed gains are large.
+  #
+  # It is important to note that this config value is global to the geoip_type. That is to say all instances of the geoip filter
+  # of the same geoip_type share the same cache. The last declared cache size will 'win'. The reason for this is that there would be no benefit
+  # to having multiple caches for different instances at different points in the pipeline, that would just increase the
+  # number of cache misses and waste memory.
+  config :lru_cache_size, :validate => :number, :default => 1000
+
   public
   def register
-    require "geoip"
```

Comment:
STORY TIME!

The reason Logstash historically put `require "some third party thing"` in the `register` method was because .... computers.

The libraries we used in Logstash (JRuby) were not always available or installed for MRI, and at the time, I used MRI to generate the docs, and because the docs generator calls `eval` on the code :smiley: those top-of-the-file `require` calls would fail because the gems weren't available. Why MRI for docs building (at the time)? Because, way back when, the only "good" markdown libraries for Ruby only worked in MRI. Uuuuugh computers.

So that's the story. `require` in `register` because of markdown and mri and jruby.

Today? I don't think these requirements are necessarily true anymore, though it would be nice to be able to build the docs _without_ needing the runtime dependencies, we can probably hack around that as needed.

+1 on, long term, moving requires to the top of the file.